scio-hdfs/src/main/scala/com/spotify/scio/hdfs/package.scala

```diff
@@ -99,6 +109,71 @@ package object hdfs {
         .map(_.getKey.datum())
     }
 
+    /**
+     * Create a new [[com.spotify.scio.values.DistCache DistCache]] instance for a file on
+     * Hadoop/HDFS.
+     *
+     * @param path to the Hadoop/HDFS artifact
+     * @param conf optional custom Hadoop configuration
+     * @param username optional Hadoop Simple Authentication remote username
+     */
+    def hadoopDistCache[F](path: String,
+                           conf: Configuration = null,
+                           username: String = null)
+                          (initFn: File => F): DistCache[F] = self.pipelineOp {
+      if (self.isTest) {
+        //TODO: if test - can the path be remote HDFS/S3 etc?
+        self.distCache(path)(initFn)
+      } else {
+        //TODO: should upload be asynchronous, blocking on context close
+        require(self.options.getStagingLocation != null,
+          "Staging directory not set - use `--stagingLocation`!")
+        require(path != null, "Artifact path can't be null")
+
+        val _conf = Option(conf).getOrElse(new Configuration())
+
+        //TODO: should we add checksums on both src and GCS to reuse uploaded artifacts?
+        val path_hash = Hashing.sha1().hashString(path, Charsets.UTF_8)
+        val target_hash = path_hash.toString.substring(0, 8)
+
+        logger.debug(s"Add '$path' (hash: '$path_hash') to dist cache")
+
+        val target_distcache =  new Path("distcache", s"$target_hash-${path.split("/").last}")
+
+        val target = new Path(self.options.getStagingLocation, target_distcache)
```

Comment:
Check if target already exists to avoid extra work?
DF does that for staging jars too.